7주차 과제

juunini opened the pull request.

juunini committed:
Add spring security

juunini committed:
Add security config class

juunini committed:
Disable csrf web security

juunini committed:
Add jwt auth filter

juunini committed:
Add jwt check in filter

juunini committed:
Add path check

juunini committed:
Add invalid token case throw exception

juunini committed:
Add auth error filter

juunini committed:
Remove intercepter

juunini committed:
Remove web config

juunini commented:
첫번 째 동영상 시청 한 부분까지 올립니다.
발목에 나사 뽑는 수술을 했는데 도저히 맨정신으로 더 버티기가 힘들군요 ㅠㅠ

johngrib commented:
> 첫번 째 동영상 시청 한 부분까지 올립니다.
> 발목에 나사 뽑는 수술을 했는데 도저히 맨정신으로 더 버티기가 힘들군요 ㅠㅠ

헉 걱정이 됩니다. ㅠㅜㅠ

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,28 @@
+        Filter authenticationFilter = new JwtAuthenticationFilter(authenticationManager(), authenticationService);
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        http
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter, JwtAuthenticationFilter.class);
```
여기 인덴트가 너무 깊어 보입니다.

johngrib reviewed (commented):
몸이 안 좋으신 와중에서도 노력하시는 모습이 대단합니다. 그러나 너무 무리 하지 않으셨으면 좋겠어요. 지금도 너무 잘 하고 계십니다!

juunini committed:
Add user auth

juunini committed:
Remove filter with path

juunini committed:
Set security annotation

juunini committed:
Add auth entry point in security config

juunini committed:
Add authorities

juunini committed:
Add config in session

juunini committed:
Remove use changeWith in password

juunini committed:
Add change password

juunini committed:
Change authenticate method

juunini committed:
Set password encoder DI

juunini committed:
Fix auth service test crash

juunini committed:
Move bean to app

johngrib reviewed (commented):
어느새 7주차가 되었어요. 이제부터 커밋 메시지 타이틀과 본문을 한국어로 작성해주시면 커밋 메시지에 대해서도 리뷰를 해 드릴게요.

johngrib reviewed (commented):
78050061f473bd8d95f287e72e7bfecbd6e227c1 어떤 의미를 갖는 애노테이션인지 커밋 메시지 본문에 설명을 추가해주셨으면 더 좋았을 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -21,7 +23,14 @@ public UserAuthentication(Long userId) {
 
     private static List<GrantedAuthority> authorities() {
-        return null;
+        List<GrantedAuthority> authorities = new ArrayList<>();
+        authorities.add(new SimpleGrantedAuthority("USER"));
+        return authorities;
```
`List.of`의 사용도 고려할 수 있을 것 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -42,4 +44,9 @@ public void destroy() {
         return !deleted && password.equals(this.password);
     }
+
+    public void changePassword(String password) {
```
이런 중요한 public 메소드에는 JavaDoc 주석을 꼭 작성해 줍시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -18,6 +18,16 @@ void changeWith() {
+        user.changePassword("TEST");
+
+        assertThat(user.getPassword()).isNotEmpty();
+        assertThat(user.getPassword()).isEqualTo("TEST");
```
`changePassword` 메소드를 호출하기 전에 `User`가 갖고 있는 패스워드가 `TEST`가 아니라는 것을 증명하는 코드가 추가로 필요해요.

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -16,4 +18,9 @@ public static void main(String[] args) {
         return DozerBeanMapperBuilder.buildDefault();
     }
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
```
`SecurityJavaConfig.java`에서 `App.java`로 `passwordEncoder`를 옮겨온 이유가 무엇인가요?

juunini replied:
강의 영상을 보면서 쭉 따라오고 있었는데 '순환참조' 를 하게되며 문제가 생기더라구요
어떤 식으로 순환참조가 되는지 관심이 생기기 시작했어요 ㅎㅎ

golang에서는 기본적으로 순환참조를 애초에 못하게 하는지라 거기에 익숙해져서
팀 내에서 코드리뷰 할 때 누군가 순환참조를 걸면 잘 잡아내는 편인데
이건 생각도 못했었거든요

johngrib replied:
이 이야기가 그대로 커밋 메시지에 있었다면 더 좋았을 거에요.

juunini committed:
createWithoutAccessToken 테스트를 describe-context-it 으로 변경하라

- http 요청별로 분기를 나누고 싶기 때문
- context를 통한 분기를 나누고 싶기 때문

juunini committed:
access token 없이 /products POST 요청을 보낼 시 unauthorized 를 응답하도록 하라

- PreAuthorize 애노테이션으로 토큰을 검증하므로 RequestAttribute로 userId를 받을 필요가 더 이상 없다.

juunini committed:
나머지 /products 의 post 요청들도 모두 describe-context-it 구조로 바꾸어라

juunini committed:
/products post 테스트의 productService mocking을 context 구문 아래로 옮겨라

juunini committed:
/products delete 요청의 테스트를 describe-context-it 구조로 변경하라

juunini committed:
/products patch 요청에 대한 테스트를 describe-context-it 구조로 변경하라

juunini committed:
/products 및 /products/{id} GET 요청의 테스트를 describe-context-it 구조로 변경하라

juunini committed:
given을 Mockito로 바꾸어라

given - throw 구문에서 에러가 나며 테스트를 실패한다. Mockito.doThrow 를 하니 문제가 해결되어 이렇게 처리하기로 했다.

juunini committed:
UserController 에서 유저 인증이 되었을 때만 patch 요청과 delete 요청이 가능하도록 수정하라

juunini committed:
ProductController 에서 AuthenticationService 를 제거하라

juunini committed:
UserControllerTest 에서 token에 따라 결과가 달라지도록 세팅하라

juunini committed:
/users POST 요청의 테스트를 describe-context-it 구조로 바꾸어라

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -265,4 +255,25 @@ void destroyWithInvalidAccessToken() throws Exception {
+            void It_respond_unauthorized() throws Exception {
+                mockMvc.perform(
+                        post("/products")
+                                .accept(MediaType.APPLICATION_JSON_UTF8)
```
`MediaType.APPLICATION_JSON_UTF8`은 Deprecated 되었습니다. 다행히 무엇으로 바꿔야 하는지는 `MediaType.APPLICATION_JSON_UTF8`의 JavaDoc을 읽어보시면 나와 있어요. 읽고 바꿔주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -275,5 +233,66 @@ void It_respond_unauthorized() throws Exception {
+            }
+        }
+
+        @Nested
+        @DisplayName("주어진 요청의 access token 이 올바를 때")
+        class Context_with_valid_access_token {
```
VALID_TOKEN을 여기에서 선언했다면 더 좋았겠지만...

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -275,5 +233,66 @@ void It_respond_unauthorized() throws Exception {
+                }
+            }
+
+            @Nested
+            @DisplayName("주어진 상품의 정보가 올바르지 않다면")
+            class Context_without_correct_product_attributes {
```
올바르지 않은 정보를 가진 상품을 이 스코프에서 정의해 주세요.

johngrib reviewed (commented):
87254c64a9b1e234529fe497472a94436bf093d0

커밋 메시지 타이틀만 읽어서는 어디를 왜 고친 것인지 알 수가 없어요. 그리고 준희님의 다른 커밋 메시지와 사뭇 다른 느낌이 듭니다.

![image](https://user-images.githubusercontent.com/1855714/111030404-665e5480-8445-11eb-9786-050370cef9e9.png)

다음과 같이 작성하면 어떨까요?

```
ProductControllerTest에서 실패하는 테스트를 고쳐라

`given - throw` 구문에서 에러가 발생하여 테스트를 실패합니다.
`Mockito.doThrow`를 사용해 문제를 해결할 수 있었습니다.
```

johngrib reviewed (commented):
875ef31f2c78352876ae994224edbc22145d036a 커밋 메시지를 더 심플하게 작성해도 좋을 것 같아요.

`UserController 에서 유저 인증이 되었을 때만 patch 요청과 delete 요청이 가능하도록 수정하라`

이걸 이렇게 해보면 어떨까요?

`인증된 사용자만 PATCH, DELETE 요청을 보낼 수 있다`

juunini committed:
UserController의 PATCH 요청의 테스트를 describe-context-it 구조로 변경하라

juunini committed:
UserController의 DELETE 요청의 테스트를 describe-context-it 구조로 변경하라

juunini committed:
UserAuthentication 의 authorities 메서드의 add 메서드 대산 List.of 를 이용하라

juunini committed:
UserAuthentication 의 커버리지를 100% 달성하라

juunini committed:
App 클래스의 커버리지 100%를 달성하라

juunini committed:
User 클래스와 Product 클래스의 커버리지 100%를 달성하라

juunini committed:
dto 클래스들의 커버리지 100%를 달성하라

- SessionResponse 클래스의 toString 메서드의 테스트 추가
- UserModificationData 클래스의 getPassword, toString 메서드의 테스트 추가
- UserRegistrationData 클래스의 toString 메서드의 테스트 추가
- UserResultData 클래스의 toString 메서드의 테스트 추가
- ProductData 클래스의 toString, id, imageUrl 메서드의 테스트 추가

juunini committed:
ControllerErrorAdvice 클래스의 커버리지 100%를 달성하라

- handleInvalidAccessTokenException 제거
- handleConstraintValidateError 제거

juunini committed:
타인의 id 에 대해 UPDATE, DELETE 시도 시 FORBIDDEN 을 응답하도록 하라

- AccessDeniedException 을 던지면 Status Forbidden 을 응답한다.
- AuthenticationService 클래스에서 jwt encode를 할 때 실제 유저의 id를 넣도록 한다.
- JwtUtil 클래스에서 userId에 주어진 id를 넣도록 한다.

juunini committed:
Forbidden을 던짐으로써 깨지는 테스트들을 고쳐라

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -221,4 +162,108 @@ void It_respond_bad_request() throws Exception {
+                void setup() {
+                    Mockito.doThrow(new UserNotFoundException(100L))
+                            .when(userService).updateUser(eq(100L), any(UserModificationData.class));
+                }
```
컨텍스트 설명문에 맞게 컨텍스트 조건을 잘 설정하신 것 같습니다. 그런데 `100L`은 `Context`와 `It`에서 중복으로 사용하고 있으니 컨텍스트 스코프에서 상수로 선언해주시는 것이 좋겠어요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -36,122 +47,238 @@
+                }
+            }
+
+            @Nested
+            @DisplayName("주어진 유저의 정보가 올바르지 않을 때")
+            class Context_without_correct_user_attributes {
+                @Test
+                @DisplayName("bad request 를 응답한다.")
+                void It_respond_bad_request() throws Exception {
+                    mockMvc.perform(
+                            patch("/users/1")
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .content("{\"name\":\"\",\"password\":\"\"}")
+                                    .header("Authorization", "Bearer " + VALID_TOKEN)
+                    )
+                            .andExpect(status().isBadRequest());
+                }
+            }
+        }
```
이렇게 했다면 컨텍스트의 의미를 더 살릴 수 있었을 거에요.
```suggestion
            @Nested
            @DisplayName("주어진 유저의 정보가 올바르지 않을 때")
            class Context_without_correct_user_attributes {
                final String givenUserData = "{\"name\":\"\",\"password\":\"\"}";

                @Test
                @DisplayName("bad request 를 응답한다.")
                void It_respond_bad_request() throws Exception {
                    mockMvc.perform(
                            patch("/users/1")
                                    .contentType(MediaType.APPLICATION_JSON)
                                    .content(givenUserData)
                                    .header("Authorization", "Bearer " + VALID_TOKEN)
                    )
                            .andExpect(status().isBadRequest());
                }
            }
        }
```

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private static List<GrantedAuthority> authorities() {
+        return new ArrayList<>(
+                List.of(new SimpleGrantedAuthority("USER"))
+        );
```
이렇게 해도 충분하지 않았을까요?
```suggestion
        return List.of(new SimpleGrantedAuthority("USER"));
```

johngrib reviewed (commented):
dc7001eae6e62f3b152e533ee7ecbbc60f774fbb
`UserAuthentication 의 커버리지를 100% 달성하라`

아마 테스트 커버리지를 의미하려 하신 것 같아요.

`UserAuthentication의 테스트 커버리지 100%를 달성하라`

![image](https://user-images.githubusercontent.com/1855714/111068089-a20f2200-850a-11eb-99a3-8a2edc9014ab.png)

johngrib reviewed (commented):
1482090dd6103c9c22f1ee50e3b1982a3ab6c629 

작업량이 많은 대형 커밋이기 때문에 이런 식의 설명은 적절한 것 같아요. 커밋 메시지를 먼저 읽으니 작업 결과를 훑어보기 좋았습니다.

![image](https://user-images.githubusercontent.com/1855714/111068151-ea2e4480-850a-11eb-854c-0ebf81c8f09c.png)

johngrib reviewed (commented):
c7e4917e2c94fff88c3c3ddced4c52b1ad7a9b67 이 커밋은 878dc24152d3b4f538543d7607ba20d729c45e4d 커밋과 합쳤(fixup)다면 더 좋았을 것 같아요.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다!
